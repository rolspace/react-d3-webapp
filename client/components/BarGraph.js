import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import CircularProgress from '@material-ui/core/CircularProgress'
import Grid from '@material-ui/core/Grid'
import GraphRenderer from '../common/bargraph'

const styles = (theme) => ({
	circle: {
		r: 5
	},
	container: {
		overflowX: 'scroll'
	},
	root: {
		width: '100%'
	},
	svg: {
		fontFamily: theme.typography.fontFamily,
		height: '75vh',
		minWidth: '700px',
		width: '100%'
	}
})

class BarGraph extends React.Component {
	constructor(props) {
		super(props)
		this.renderGraph = this.renderGraph.bind(this)
	}

	componentDidUpdate() {
		this.renderGraph()
	}

	renderGraph() {
		const node = this.node;

		const data = {
			sets: this.props.data.sets,
			xAxis: this.props.data.xAxis,
			yAxis: this.props.data.yAxis,
			xAxisLabel: this.props.data.xAxisLabel,
			yAxisLabel: this.props.data.yAxisLabel,
			height: 500,
			width: 800
		}

		const graphRenderer = new GraphRenderer(node, data)
		graphRenderer.renderGraph()
	}

	render() {
		const { classes } = this.props

		return (
			<div>
				<Grid container className={classes.container}>
					<Grid item xs={12}>
						{this.props.data.isLoading ?
							<CircularProgress size='100%' thickness={1} classes={{svg: classes.svg, circleIndeterminate: classes.circle}} />
							:
							<svg ref={node => this.node = node} className={classes.svg} viewBox='0 0 800 500'></svg>
						}
					</Grid>
				</Grid>
			</div>
		)
	}
}

BarGraph.propTypes = {
	classes: PropTypes.object.isRequired,
	data: PropTypes.object.isRequired,
}

export default withStyles(styles)(BarGraph)