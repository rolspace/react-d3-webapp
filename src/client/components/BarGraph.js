import CircularProgress from '@material-ui/core/CircularProgress'
import Grid from '@material-ui/core/Grid'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import React, { useEffect, useRef } from 'react'
import { renderBarGraph } from '../common/bargraph'

const styles = (theme) => ({
  circle: {
    r: 5,
  },
  container: {
    overflowX: 'scroll',
  },
  root: {
    width: '100%',
  },
  svg: {
    fontFamily: theme.typography.fontFamily,
    height: '73vh',
    minWidth: '700px',
    width: '100%',
  },
  circleRoot: {
    left: '45%',
    position: 'absolute',
    top: '50%',
  },
})

const BarGraph = ({ graphData, classes }) => {
  const { sets, colors, xAxis, xAxisLabel, yAxis, yAxisLabel, isLoading } =
    graphData

  const svgRef = useRef(null)

  useEffect(() => {
    const renderGraph = () => {
      const { current: node } = svgRef

      if (node && sets) {
        renderBarGraph(node, {
          sets,
          colors,
          xAxis,
          xAxisLabel,
          yAxis,
          yAxisLabel,
          height: 500,
          width: 800,
        })
      }
    }

    renderGraph()
  }, [graphData])

  return (
    <div>
      <Grid container className={classes.container}>
        <Grid item xs={12} style={{ height: '75vh' }}>
          {isLoading ? (
            <CircularProgress classes={{ root: classes.circleRoot }} />
          ) : (
            <svg
              ref={svgRef}
              className={classes.svg}
              viewBox="0 0 800 500"></svg>
          )}
        </Grid>
      </Grid>
    </div>
  )
}

BarGraph.propTypes = {
  classes: PropTypes.object.isRequired,
  graphData: PropTypes.object.isRequired,
}

export default withStyles(styles)(BarGraph)
