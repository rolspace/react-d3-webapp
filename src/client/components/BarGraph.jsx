import CircularProgress from '@mui/material/CircularProgress'
import Grid from '@mui/material/Grid'
import { styled } from '@mui/material/styles'
import PropTypes from 'prop-types'
import React, { useEffect, useRef } from 'react'
import { renderBarGraph } from '../services/bargraph.js'

const PREFIX = 'BarGraph'

const classes = {
  circle: `${PREFIX}-circle`,
  container: `${PREFIX}-container`,
  root: `${PREFIX}-root`,
  svg: `${PREFIX}-svg`,
  circleRoot: `${PREFIX}-circleRoot`,
}

const Root = styled('div')((
  {
    theme,
  },
) => ({
  [`& .${classes.circle}`]: {
    r: 5,
  },

  [`& .${classes.container}`]: {
    overflowX: 'scroll',
  },

  [`& .${classes.root}`]: {
    width: '100%',
  },

  [`& .${classes.svg}`]: {
    fontFamily: theme?.typography?.fontFamily,
    height: '73vh',
    minWidth: '700px',
    width: '100%',
  },

  [`& .${classes.circleRoot}`]: {
    left: '45%',
    position: 'absolute',
    top: '50%',
  },
}))

const BarGraph = ({ graphData }) => {
  const { sets, colors, loading, xAxis, xAxisLabel, yAxis, yAxisLabel } =
    graphData

  const svgRef = useRef(null)

  useEffect(() => {
    const renderGraph = () => {
      const { current: node } = svgRef

      if (node && sets) {
        renderBarGraph(node, {
          sets,
          colors,
          xAxis,
          xAxisLabel,
          yAxis,
          yAxisLabel,
          height: 500,
          width: 800,
        })
      }
    }

    renderGraph()
  }, [graphData])

  return (
    <Root>
      <Grid container className={classes.container}>
        <Grid size={12} style={{ height: '75vh' }}>
          {loading === 'pending' ? (
            <CircularProgress classes={{ root: classes.circleRoot }} />
          ) : (
            <svg
              ref={svgRef}
              className={classes.svg}
              viewBox="0 0 800 500"
            />
          )}
        </Grid>
      </Grid>
    </Root>
  )
}

BarGraph.propTypes = {
  classes: PropTypes.object.isRequired,
  graphData: PropTypes.object.isRequired,
}

export default BarGraph
