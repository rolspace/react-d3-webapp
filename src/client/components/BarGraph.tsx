import CircularProgress from '@mui/material/CircularProgress'
import Grid from '@mui/material/Grid'
import { styled, Theme } from '@mui/material/styles'
import { FC, useEffect, useRef } from 'react'
import { renderBarGraph } from '../services/barGraph'
import { BarGraphDataSets } from '../types/barGraph.types'

const PREFIX = 'BarGraph'

const classes = {
  circle: `${PREFIX}-circle`,
  container: `${PREFIX}-container`,
  root: `${PREFIX}-root`,
  svg: `${PREFIX}-svg`,
  circleRoot: `${PREFIX}-circleRoot`,
}

const Root = styled('div')<{ theme?: Theme }>(({ theme }) => ({
  [`& .${classes.circle}`]: {
    r: 5,
  },

  [`& .${classes.container}`]: {
    overflowX: 'scroll',
  },

  [`& .${classes.root}`]: {
    width: '100%',
  },

  [`& .${classes.svg}`]: {
    fontFamily: theme?.typography?.fontFamily,
    height: '73vh',
    minWidth: '700px',
    width: '100%',
  },

  [`& .${classes.circleRoot}`]: {
    left: '45%',
    position: 'absolute',
    top: '50%',
  },
}))

interface BarGraphProps {
  graphData: BarGraphDataSets;
}

const BarGraph: FC<BarGraphProps> = ({ graphData }) => {
  const { sets, colors, loading, xAxis, xAxisLabel, yAxis, yAxisLabel } =
    graphData

  const svgRef = useRef<SVGSVGElement>(null)

  useEffect(() => {
    const renderGraph = () => {
      const { current: svgElement } = svgRef

      if (svgElement && sets) {
        renderBarGraph(svgElement, {
          sets,
          xAxis,
          yAxis,
        }, {
            colors,
            xAxisLabel,
            yAxisLabel,
            height: 500,
            width: 800,
        })
      }
    }

    renderGraph()
  }, [graphData])

  return (
    <Root>
      <Grid container className={classes.container}>
        <Grid size={12} style={{ height: '75vh' }}>
          {loading === 'pending' ? (
            <CircularProgress classes={{ root: classes.circleRoot }} />
          ) : (
            <svg
              ref={svgRef}
              className={classes.svg}
              viewBox="0 0 800 500"
            />
          )}
        </Grid>
      </Grid>
    </Root>
  )
}

export default BarGraph
