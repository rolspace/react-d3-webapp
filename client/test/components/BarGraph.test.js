import React from 'react'
import renderer from 'react-test-renderer'
import BarGraph from '../../components/BarGraph'

describe('Components: BarGraph component', () => {
  it('renders correctly', () => {
    const data = {
      sets: [],
      xAxis: 'xAxis',
			yAxis: 'yAxis',
			xAxisLabel: 'xAxisLabel',
			yAxisLabel: 'yAxisLabel'
    }

    const tree = renderer.create(<BarGraph data={data} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})