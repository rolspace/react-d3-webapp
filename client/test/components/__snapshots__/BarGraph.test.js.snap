// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: BarGraph component renders correctly 1`] = `
<div>
  <div
    className="MuiGrid-container-5 BarGraph-container-2"
  >
    <div
      className="MuiGrid-item-6 MuiGrid-grid-xs-12-45"
    >
      <svg
        className="BarGraph-svg-4"
        viewBox="0 0 800 500"
      />
    </div>
  </div>
</div>
`;
